File: client/src/api/updates.ts
/**
 * Update Service API client for BaluHost
 */
import { apiClient } from '../lib/api';

// Update status types
export type UpdateStatus =
  | 'pending'
  | 'checking'
  | 'downloading'
  | 'backing_up'
  | 'installing'
  | 'migrating'
  | 'restarting'
  | 'health_check'
  | 'completed'
  | 'failed'
  | 'rolled_back'
  | 'cancelled';

export type UpdateChannel = 'stable' | 'unstable';

// Version information
export interface VersionInfo {
  version: string;
  commit: string;
  commit_short: string;
  tag: string | null;
  date: string | null;
  is_dev_build: boolean;
}

// Changelog entry
export interface ChangelogEntry {
  version: string;
  date: string | null;
  changes: string[];
  breaking_changes: string[];
  is_prerelease: boolean;
}

// Check response
export interface UpdateCheckResponse {
  update_available: boolean;
  current_version: VersionInfo;
  latest_version: VersionInfo | null;
  changelog: ChangelogEntry[];
  channel: UpdateChannel;
  last_checked: string | null;
  blockers: string[];
  can_update: boolean;
  dev_version_available: boolean;
  dev_version: VersionInfo | null;
  dev_commits_ahead: number | null;
  dev_commits: CommitInfo[];
}

// Start request/response
export interface UpdateStartRequest {
  target_version?: string | null;
  skip_backup?: boolean;
  force?: boolean;
}

export interface UpdateStartResponse {
  success: boolean;
  update_id: number | null;
  message: string;
  blockers: string[];
}

// Progress response
export interface UpdateProgressResponse {
  update_id: number;
  status: UpdateStatus;
  progress_percent: number;
  current_step: string | null;
  started_at: string;
  estimated_remaining: number | null;
  from_version: string;
  to_version: string;
  error_message: string | null;
  can_rollback: boolean;
}

// History entry
export interface UpdateHistoryEntry {
  id: number;
  from_version: string;
  to_version: string;
  channel: UpdateChannel;
  from_commit: string;
  to_commit: string;
  started_at: string;
  completed_at: string | null;
  duration_seconds: number | null;
  status: UpdateStatus;
  error_message: string | null;
  rollback_commit: string | null;
  user_id: number | null;
  can_rollback: boolean;
}

export interface UpdateHistoryResponse {
  updates: UpdateHistoryEntry[];
  total: number;
  page: number;
  page_size: number;
}

// Rollback request/response
export interface RollbackRequest {
  target_update_id?: number | null;
  target_commit?: string | null;
  restore_backup?: boolean;
}

export interface RollbackResponse {
  success: boolean;
  message: string;
  update_id: number | null;
  rolled_back_to: string | null;
}

// Config
export interface UpdateConfig {
  id: number;
  auto_check_enabled: boolean;
  check_interval_hours: number;
  channel: UpdateChannel;
  auto_backup_before_update: boolean;
  require_healthy_services: boolean;
  auto_update_enabled: boolean;
  auto_update_window_start: string | null;
  auto_update_window_end: string | null;
  last_check_at: string | null;
  last_available_version: string | null;
  updated_at: string;
  updated_by: number | null;
}

export interface UpdateConfigUpdate {
  auto_check_enabled?: boolean;
  check_interval_hours?: number;
  channel?: UpdateChannel;
  auto_backup_before_update?: boolean;
  require_healthy_services?: boolean;
  auto_update_enabled?: boolean;
  auto_update_window_start?: string | null;
  auto_update_window_end?: string | null;
}

// Release notes
export interface ReleaseNoteCategory {
  name: string;
  icon: string;
  changes: string[];
}

export interface ReleaseNotesResponse {
  version: string;
  previous_version: string | null;
  date: string | null;
  categories: ReleaseNoteCategory[];
}

// Commit history types (Versions tab)
export interface CommitInfo {
  hash: string;
  hash_short: string;
  message: string;
  date: string;
  author: string;
  type: string | null;
  scope: string | null;
}

export interface VersionGroup {
  tag: string | null;
  version: string;
  date: string | null;
  commit_count: number;
  commits: CommitInfo[];
}

export interface CommitHistoryResponse {
  total_commits: number;
  groups: VersionGroup[];
}

export interface DiffFile {
  path: string;
  status: string;
  additions: number;
  deletions: number;
}

export interface CommitDiffResponse {
  hash: string;
  hash_short: string;
  message: string;
  date: string;
  author: string;
  stats: string;
  files: DiffFile[];
  diff: string;
}

// Release list types
export interface ReleaseInfo {
  tag: string;
  version: string;
  date: string | null;
  is_prerelease: boolean;
  commit_short: string;
}

export interface ReleaseListResponse {
  releases: ReleaseInfo[];
  total: number;
}

// Version history types (Version Tracker)
export interface VersionHistoryEntry {
  id: number;
  version: string;
  git_commit: string;
  git_commit_short: string;
  git_branch: string | null;
  python_version: string | null;
  first_seen: string;
  last_seen: string;
  times_started: number;
}

export interface VersionHistoryResponse {
  versions: VersionHistoryEntry[];
  total: number;
  current_version: string;
  current_commit: string;
}

// API Functions

/**
 * Get current version (public endpoint, no auth required)
 */
export async function getPublicVersion(): Promise<VersionInfo> {
  const response = await apiClient.get<VersionInfo>('/api/updates/version');
  return response.data;
}

/**
 * Get release notes for the current version (public endpoint, no auth required)
 */
export async function getReleaseNotes(): Promise<ReleaseNotesResponse> {
  const response = await apiClient.get<ReleaseNotesResponse>('/api/updates/release-notes');
  return response.data;
}

/**
 * Check for available updates
 */
export async function checkForUpdates(): Promise<UpdateCheckResponse> {
  const response = await apiClient.get<UpdateCheckResponse>('/api/updates/check');
  return response.data;
}

/**
 * Start an update
 */
export async function startUpdate(request?: UpdateStartRequest): Promise<UpdateStartResponse> {
  const response = await apiClient.post<UpdateStartResponse>('/api/updates/start', request ?? {});
  return response.data;
}

/**
 * Get update progress by ID
 */
export async function getUpdateProgress(updateId: number): Promise<UpdateProgressResponse> {
  const response = await apiClient.get<UpdateProgressResponse>(`/api/updates/progress/${updateId}`);
  return response.data;
}

/**
 * Get the currently running update (if any)
 */
export async function getCurrentUpdate(): Promise<UpdateProgressResponse | null> {
  const response = await apiClient.get<UpdateProgressResponse | null>('/api/updates/current');
  return response.data;
}

/**
 * Rollback to a previous version
 */
export async function rollbackUpdate(request?: RollbackRequest): Promise<RollbackResponse> {
  const response = await apiClient.post<RollbackResponse>('/api/updates/rollback', request ?? {});
  return response.data;
}

/**
 * Get update history
 */
export async function getUpdateHistory(
  options: { page?: number; page_size?: number } = {}
): Promise<UpdateHistoryResponse> {
  const response = await apiClient.get<UpdateHistoryResponse>('/api/updates/history', {
    params: {
      page: options.page ?? 1,
      page_size: options.page_size ?? 20,
    },
  });
  return response.data;
}

/**
 * Get update configuration
 */
export async function getUpdateConfig(): Promise<UpdateConfig> {
  const response = await apiClient.get<UpdateConfig>('/api/updates/config');
  return response.data;
}

/**
 * Update configuration
 */
export async function updateConfig(config: UpdateConfigUpdate): Promise<UpdateConfig> {
  const response = await apiClient.put<UpdateConfig>('/api/updates/config', config);
  return response.data;
}

/**
 * Get full commit history grouped by version tags
 */
export async function getCommitHistory(): Promise<CommitHistoryResponse> {
  const response = await apiClient.get<CommitHistoryResponse>('/api/updates/commits');
  return response.data;
}

/**
 * Get diff details for a specific commit
 */
export async function getCommitDiff(commitHash: string): Promise<CommitDiffResponse> {
  const response = await apiClient.get<CommitDiffResponse>(`/api/updates/commits/${commitHash}/diff`);
  return response.data;
}

/**
 * Get list of all releases (git tags)
 */
export async function getAllReleases(): Promise<ReleaseListResponse> {
  const response = await apiClient.get<ReleaseListResponse>('/api/updates/releases');
  return response.data;
}

/**
 * Get version history (all versions that have ever run on this system)
 */
export async function getVersionHistory(): Promise<VersionHistoryResponse> {
  const response = await apiClient.get<VersionHistoryResponse>('/api/updates/version-history');
  return response.data;
}

// Helper functions

/**
 * Get status display info (color, icon, label)
 */
export function getStatusInfo(status: UpdateStatus): {
  color: string;
  bgColor: string;
  icon: string;
  label: string;
} {
  switch (status) {
    case 'pending':
      return {
        color: 'text-slate-400',
        bgColor: 'bg-slate-500/20',
        icon: '⏳',
        label: 'Pending',
      };
    case 'checking':
      return {
        color: 'text-sky-400',
        bgColor: 'bg-sky-500/20',
        icon: '🔍',
        label: 'Checking',
      };
    case 'downloading':
      return {
        color: 'text-sky-400',
        bgColor: 'bg-sky-500/20',
        icon: '⬇️',
        label: 'Downloading',
      };
    case 'backing_up':
      return {
        color: 'text-amber-400',
        bgColor: 'bg-amber-500/20',
        icon: '💾',
        label: 'Creating Backup',
      };
    case 'installing':
      return {
        color: 'text-sky-400',
        bgColor: 'bg-sky-500/20',
        icon: '📦',
        label: 'Installing',
      };
    case 'migrating':
      return {
        color: 'text-purple-400',
        bgColor: 'bg-purple-500/20',
        icon: '🔄',
        label: 'Migrating Database',
      };
    case 'restarting':
      return {
        color: 'text-amber-400',
        bgColor: 'bg-amber-500/20',
        icon: '🔄',
        label: 'Restarting Services',
      };
    case 'health_check':
      return {
        color: 'text-sky-400',
        bgColor: 'bg-sky-500/20',
        icon: '🏥',
        label: 'Health Check',
      };
    case 'completed':
      return {
        color: 'text-emerald-400',
        bgColor: 'bg-emerald-500/20',
        icon: '✅',
        label: 'Completed',
      };
    case 'failed':
      return {
        color: 'text-rose-400',
        bgColor: 'bg-rose-500/20',
        icon: '❌',
        label: 'Failed',
      };
    case 'rolled_back':
      return {
        color: 'text-amber-400',
        bgColor: 'bg-amber-500/20',
        icon: '↩️',
        label: 'Rolled Back',
      };
    case 'cancelled':
      return {
        color: 'text-slate-400',
        bgColor: 'bg-slate-500/20',
        icon: '🚫',
        label: 'Cancelled',
      };
    default:
      return {
        color: 'text-slate-400',
        bgColor: 'bg-slate-500/20',
        icon: '❓',
        label: status,
      };
  }
}

/**
 * Format duration in seconds to human readable string
 */
export function formatDuration(seconds: number | null): string {
  if (seconds === null) return '-';
  if (seconds < 60) return `${seconds}s`;
  if (seconds < 3600) return `${Math.floor(seconds / 60)}m ${seconds % 60}s`;
  const hours = Math.floor(seconds / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);
  return `${hours}h ${minutes}m`;
}

/**
 * Check if an update status indicates it's in progress
 */
export function isUpdateInProgress(status: UpdateStatus): boolean {
  return [
    'pending',
    'checking',
    'downloading',
    'backing_up',
    'installing',
    'migrating',
    'restarting',
    'health_check',
  ].includes(status);
}

/**
 * Get channel display info
 */
export function getChannelInfo(channel: UpdateChannel): {
  label: string;
  color: string;
  description: string;
} {
  switch (channel) {
    case 'stable':
      return {
        label: 'Stable',
        color: 'text-emerald-400',
        description: 'Production-ready releases',
      };
    case 'unstable':
      return {
        label: 'Unstable',
        color: 'text-amber-400',
        description: 'Pre-release versions with new features',
      };
    default:
      return {
        label: channel,
        color: 'text-slate-400',
        description: '',
      };
  }
}

File: backend/app/api/routes/updates.py
"""Update service API endpoints."""
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Request, Response, status, Query
from sqlalchemy.orm import Session

from app.api import deps
from app.core.rate_limiter import limiter, user_limiter, get_limit
from app.core.database import get_db
from app.schemas.user import UserPublic
from app.schemas.update import (
    UpdateCheckResponse,
    UpdateStartRequest,
    UpdateStartResponse,
    UpdateProgressResponse,
    UpdateHistoryResponse,
    RollbackRequest,
    RollbackResponse,
    UpdateConfigResponse,
    UpdateConfigUpdate,
    VersionInfo,
    ReleaseNotesResponse,
)
from app.services.update_service import get_update_service, get_update_backend
from app.services.audit_logger_db import get_audit_logger_db

router = APIRouter(prefix="/updates", tags=["updates"])


@router.get("/version", response_model=VersionInfo)
@limiter.limit(get_limit("admin_operations"))
async def get_public_version(
    request: Request, response: Response,
) -> VersionInfo:
    """Get current version (public endpoint, no auth required).

    Returns the current installed version information including
    version string, commit hash, and tag.
    """
    backend = get_update_backend()
    return await backend.get_current_version()


@router.get("/release-notes", response_model=ReleaseNotesResponse)
@limiter.limit(get_limit("admin_operations"))
async def get_release_notes(
    request: Request, response: Response,
) -> ReleaseNotesResponse:
    """Get release notes for the current version (public endpoint, no auth required).

    Returns categorized changes between the previous and current version tag.
    """
    backend = get_update_backend()
    return await backend.get_release_notes()


@router.get("/check", response_model=UpdateCheckResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def check_for_updates(
    request: Request, response: Response,
    current_user: UserPublic = Depends(deps.get_current_admin),
    db: Session = Depends(get_db),
) -> UpdateCheckResponse:
    """Check if updates are available.

    Returns current version, latest available version, changelog,
    and any blockers that would prevent updating.

    Requires admin privileges.
    """
    service = get_update_service(db)
    audit_logger = get_audit_logger_db()

    result = await service.check_for_updates()

    audit_logger.log_event(
        event_type="UPDATE",
        action="check_updates",
        user=current_user.username,
        resource="system",
        success=True,
        details={
            "current_version": result.current_version.version,
            "update_available": result.update_available,
            "latest_version": result.latest_version.version if result.latest_version else None,
        },
        db=db,
    )

    return result


@router.post("/start", response_model=UpdateStartResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def start_update(
    request: Request, response: Response,
    body: Optional[UpdateStartRequest] = None,
    current_user: UserPublic = Depends(deps.get_current_admin),
    db: Session = Depends(get_db),
) -> UpdateStartResponse:
    """Start the update process.

    Initiates an update to the latest available version (or specified version).
    The update runs in the background and progress can be monitored via
    the /progress/{id} endpoint.

    Requires admin privileges.
    """
    service = get_update_service(db)
    audit_logger = get_audit_logger_db()

    req = body or UpdateStartRequest()

    result = await service.start_update(
        user_id=current_user.id,
        target_version=req.target_version,
        skip_backup=req.skip_backup,
        force=req.force,
    )

    audit_logger.log_event(
        event_type="UPDATE",
        action="start_update",
        user=current_user.username,
        resource="system",
        success=result.success,
        details={
            "update_id": result.update_id,
            "target_version": req.target_version,
            "skip_backup": req.skip_backup,
            "force": req.force,
            "message": result.message,
            "blockers": result.blockers,
        },
        db=db,
    )

    if not result.success and result.blockers:
        # Return 409 Conflict if blocked
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={
                "message": result.message,
                "blockers": result.blockers,
            }
        )

    return result


@router.get("/progress/{update_id}", response_model=UpdateProgressResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def get_update_progress(
    request: Request, response: Response,
    update_id: int,
    current_user: UserPublic = Depends(deps.get_current_admin),
    db: Session = Depends(get_db),
) -> UpdateProgressResponse:
    """Get progress of an update.

    Returns current status, progress percentage, and current step
    for the specified update.

    Requires admin privileges.
    """
    service = get_update_service(db)

    result = service.get_update_progress(update_id)

    if not result:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Update {update_id} not found",
        )

    return result


@router.get("/current", response_model=Optional[UpdateProgressResponse])
@user_limiter.limit(get_limit("admin_operations"))
async def get_current_update(
    request: Request, response: Response,
    current_user: UserPublic = Depends(deps.get_current_admin),
    db: Session = Depends(get_db),
) -> Optional[UpdateProgressResponse]:
    """Get the currently running update, if any.

    Returns None if no update is in progress.

    Requires admin privileges.
    """
    from app.models.update_history import UpdateHistory, UpdateStatus

    # Find any running update
    running = (
        db.query(UpdateHistory)
        .filter(UpdateHistory.status.in_([
            UpdateStatus.PENDING.value,
            UpdateStatus.CHECKING.value,
            UpdateStatus.DOWNLOADING.value,
            UpdateStatus.BACKING_UP.value,
            UpdateStatus.INSTALLING.value,
            UpdateStatus.MIGRATING.value,
            UpdateStatus.RESTARTING.value,
            UpdateStatus.HEALTH_CHECK.value,
        ]))
        .first()
    )

    if not running:
        return None

    service = get_update_service(db)
    return service.get_update_progress(running.id)


@router.post("/rollback", response_model=RollbackResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def rollback_update(
    request: Request, response: Response,
    body: Optional[RollbackRequest] = None,
    current_user: UserPublic = Depends(deps.get_current_admin),
    db: Session = Depends(get_db),
) -> RollbackResponse:
    """Rollback to a previous version.

    Can specify a target update ID or commit to rollback to.
    If not specified, rolls back to the previous successful update.

    Requires admin privileges.
    """
    service = get_update_service(db)
    audit_logger = get_audit_logger_db()

    req = body or RollbackRequest()

    result = await service.rollback(req, current_user.id)

    audit_logger.log_event(
        event_type="UPDATE",
        action="rollback",
        user=current_user.username,
        resource="system",
        success=result.success,
        details={
            "target_update_id": req.target_update_id,
            "target_commit": req.target_commit,
            "restore_backup": req.restore_backup,
            "rolled_back_to": result.rolled_back_to,
            "message": result.message,
        },
        db=db,
    )

    if not result.success:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=result.message,
        )

    return result


@router.get("/history", response_model=UpdateHistoryResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def get_update_history(
    request: Request, response: Response,
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(20, ge=1, le=100, description="Items per page"),
    current_user: UserPublic = Depends(deps.get_current_admin),
    db: Session = Depends(get_db),
) -> UpdateHistoryResponse:
    """Get update history.

    Returns paginated list of past updates with status and details.

    Requires admin privileges.
    """
    service = get_update_service(db)
    return service.get_history(page=page, page_size=page_size)


@router.get("/config", response_model=UpdateConfigResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def get_update_config(
    request: Request, response: Response,
    current_user: UserPublic = Depends(deps.get_current_admin),
    db: Session = Depends(get_db),
) -> UpdateConfigResponse:
    """Get update service configuration.

    Returns auto-check settings, update channel, backup preferences, etc.

    Requires admin privileges.
    """
    service = get_update_service(db)
    return service.get_config()


@router.put("/config", response_model=UpdateConfigResponse)
@user_limiter.limit(get_limit("admin_operations"))
async def update_config(
    request: Request, response: Response,
    config: UpdateConfigUpdate,
    current_user: UserPublic = Depends(deps.get_current_admin),
    db: Session = Depends(get_db),
) -> UpdateConfigResponse:
    """Update service configuration.

    Allows modifying auto-check settings, update channel, etc.

    Requires admin privileges.
    """
    service = get_update_service(db)
    audit_logger = get_audit_logger_db()

    result = service.update_config(config, current_user.id)

    audit_logger.log_event(
        event_type="UPDATE",
        action="update_config",
        user=current_user.username,
        resource="update_config",
        success=True,
        details=config.model_dump(exclude_unset=True),
        db=db,
    )

    return result
